torch/utils/_sympy/functions.py

```diff
@@ -446,14 +478,23 @@ class PowByNatural(sympy.Function):
 
     @classmethod
     def eval(cls, base, exp):
-        if isinstance(base, sympy.Number) and isinstance(exp, sympy.Number):
-            return sympy.Integer(safe_pow(base, exp))
+        if isinstance(base, sympy.Integer) and isinstance(exp, sympy.Integer):
+            r = safe_pow(base, exp)
+            if r in (-int_oo, int_oo):
+                return r
+            return sympy.Integer(r)
         if isinstance(exp, sympy.Integer):
             # Translate power into iterated multiplication
```

Comment:
We don't want to do exponentiation by squaring?

Reply:
Hmm. Well, we'd have to add a square function, cuz sympy doesn't have one built in.

Reply:
Now, I mean, we can do `x * x` for squaring. My point here is that we can have a logarithmic number of multiplications, rather than a linear one

Reply:
Maybe we should just directly call Pow 🤔 

```
>>> import sympy
>>> from sympy.abc import x
>>> x2 = x * x
>>> x3 = x2 * x2
>>> x4 = x3 * x3
>>> x4
x**8
>>> type(x4)
<class 'sympy.core.power.Pow'>
```